tasks/upload_finisher.py

```diff
@@ -207,13 +208,47 @@ async def finish_reports_processing(
         else:
             commit.state = "skipped"
 
+        if self.should_clean_labels_index(commit_yaml, processing_results):
+            task = self.app.tasks[clean_labels_index_task_name].apply_async(
+                kwargs={
+                    "repoid": repoid,
+                    "commitid": commitid,
+                    "report_code": report_code,
+                },
+            )
+            log.info(
+                "Scheduling clean_labels_index task",
+                extra=dict(
+                    repoid=repoid,
+                    commit=commitid,
+                    clean_labels_index_task_id=task.id,
+                    parent_task=self.request.parent_id,
+                ),
+            )
+
         if checkpoints:
             checkpoints.log(UploadFlow.PROCESSING_COMPLETE)
             if not notifications_called:
                 checkpoints.log(UploadFlow.SKIPPING_NOTIFICATION)
 
         return {"notifications_called": notifications_called}
 
+    def should_clean_labels_index(self, commit_yaml: UserYaml, processing_results):
+        """Returns True if any of the successful processings was uploaded using a flag
+        that implies labels were uploaded with the report.
+        """
+        actual_processing_results = processing_results.get("processings_so_far", [])
+        for result in actual_processing_results:
+            if result["successful"]:
+                flags_str: str = result.get("arguments", {}).get("flags", "")
+                if flags_str:
+                    individual_flags = flags_str.split(",")
+                    for flag in individual_flags:
+                        config = commit_yaml.get_flag_configuration(flag)
+                        if config and config.get("carryforward_mode") == "labels":
+                            return True
+        return False
```

Comment:
i try to avoid needing this many levels of indentation, it can make logic hard to follow. can't always do it without perf consequences, but i think here perf doesn't matter

imo the code would be tidier if it broke the logic into three nested sub-checks that we could give names to:
- we need to clean the **label index** if any of our **processing results** need to be cleaned
- a **processing result** needs to be cleaned if a) it's successful, and b) any of **its flags** need to be cleaned
- a **flag** needs to be cleaned if a) it has a yaml config, and b) that config's mode is "labels"

```
def should_clean_labels_index(self, commit_yaml, processing_results):
    def should_clean_for_flag(flag):
        config = commit_yaml.get_flag_configuration(flag)
        return config and config.get("carryforward_mode", "") == "labels"

    def should_clean_for_processing_result(result):
        args = results.get("arguments", {})
        flags_str = args.get("flags", "")
        flags = flags_str.split(",") if flags_str else []
        return result["successful"] and any(map(flags, should_clean_for_flag))

    actual_processing_results = processing_results.get("processings_so_far", [])
    return any(map(actual_processing_results, should_clean_for_processing_result))
```